### ISSUE TITLE
Allow ignoring of error codes in execToString

### ISSUE THREAD

**Issue by Brunni** [XITASO/devcmd#47]
I want to call `docker inspect myContainer` and use the returned JSON result.

This works fine.
When the container does not exist I get an empty array but also an errorcode != 0.

I would like to have an option "ignoreErrorCode" or "returnErrorCode". So that I can handle the error code on my own (or not handle it). Example can be seen here in Jenkins: https://www.jenkins.io/doc/pipeline/steps/workflow-durable-task-step/#sh-shell-script

I think it needs to be added here:
https://github.com/XITASO/devcmd/blob/d34cd8bb986b58991d1947fce6783cc5b64730ea/packages/devcmd/src/process/ProcessExecutor.ts#L146

I would also prefer, if I can decide whether the output is printed to the console or not.

**Comment by lehnerpat** [comment]
Hi @Brunni, thanks for suggesting this! It's really valuable to get feedback from users about how well DevCmd's behavior is working for them (or not).

> I would like to have an option "ignoreErrorCode" or "returnErrorCode". 

I think it makes sense to allow developers to specify they don't want an error to be thrown when a process exits with a non-zero status code. For the `exec*` methods that run a single command, this is pretty clear to me, but I'd like to put some thought into how this affects `execParallel`. Feel free to post your thoughts about what behavior you'd expect here.

As a temporary workaround, I think in your case you could surround your `execToString` call with a try-catch block. 
DevCmd's own commands do this for certain Docker commands that exit non-zero when the container or image doesn't exist, for example:

https://github.com/XITASO/devcmd/blob/d34cd8bb986b58991d1947fce6783cc5b64730ea/dev_cmds/run-integration-tests.ts#L47-L49

> I would also prefer, if I can decide whether the output is printed to the console or not.

By this, which output exactly do you mean?
This seems separate from the exit code discussion, so it would be great if you could split this into a separate issue :)

**Comment by Brunni** [comment]
Hi. I haven't used execParallel that much, yet. I was using manual Promise.all(...) to run things in parallel as more logic was necessary.

Concerning the stdout output: I'm currently using the try-catch-workaround, but this still prints a red error message to the console, which might lead to confusion. So this is even more a nice2have feature.

**Comment by lehnerpat** [comment]
Thanks for adding more detail :)

For posterity, I've assembled an MWE:

Given a command script called `try-fail.ts` with the following content:
```ts
import { execPiped, runAsyncMain } from "devcmd";

async function main(): Promise<void> {
  try {
    await execPiped({ command: "bash", args: ["-c", "exit 2"] });
  } catch {}
}

runAsyncMain(main);
```

Running it still prints a report about the non-zero exit code in red:
![image](https://user-images.githubusercontent.com/1099818/147495575-d678ec30-6231-4b6d-a2b6-516a9ba82eae.png)

----

I think the following two changes include what you asked about and would give you the behavior you would like:
1. Add an option 'nonZeroExitCodeHandling' to the `exec*` methods whose type is a new enum.
  * If 'nonZeroExitCodeHandling' is `printErrorAndThrow` (default), the behavior remains as it is now: DevCmd prints a red error message and throws and error.
  * If 'nonZeroExitCodeHandling' is `printNoticeAndReturn`, DevCmd prints the message about the non-zero exit code in a "notice" style instead (the same style that's used for the "Starting process" message) and the exit code value is returned from the `exec*` method without throwing an error.
2. Add a boolean option 'suppressNotices' to the `exec*` methods:
  * If 'suppressNotices' is `false` (default), the behavior remains as it is now: DevCmd prints notices about the starting and finishing of the process.
  * If 'suppressNotices' is `true`, DevCmd does not print these notices. The process's output is still handled as before (e.g. `execPiped` still forwards it and `execToString` still collects it).

@Brunni, what do you think?

PS: Since we don't really have any API documentation yet, this might be a very hidden feature, but instead of using the `exec*` methods exported from DevCmd, you can also instantiate `ProcessExecutor` on your own, passing it a console-like object that's used for printing output. With this, you can also suppress all output, like some of our unit tests do, see `nullConsole` in [packages/devcmd/src/process/process.test.ts](https://github.com/XITASO/devcmd/blob/d34cd8bb986b58991d1947fce6783cc5b64730ea/packages/devcmd/src/process/process.test.ts).

**Comment by PatrickLehnerXI** [comment]
@Brunni, we just released version 0.0.10 of the devcmd package, which includes the feature you requested here (i.e. #48 and #53). Please try it to see if this solves your issue :)

I'm closing this issue for now. If something's not working, feel free to create a new issue!
